File: patch.js
const fs = require('fs');
let codeGen = fs.readFileSync('../phpurs/phpurs/src/Phpurs/CodeGen.purs', 'utf8');

codeGen = codeGen.replace(/freeVars :: TcoExpr -> Set String\nfreeVars \(TcoExpr \(TcoAnalysis \{ freeVars: fvs \}\) _\) = fvs/,
  `freeVars :: TcoExpr -> Set String\nfreeVars (TcoExpr (TcoAnalysis { freeVars: fvs }) _) = fvs\n\nisEffectNode :: TcoExpr -> Boolean\nisEffectNode (TcoExpr _ syntax) = case syntax of\n  EffectBind _ _ _ _ -> true\n  EffectPure _ -> true\n  EffectDefer _ -> false\n  PrimEffect _ -> true\n  UncurriedEffectApp _ _ -> true\n  Let _ _ _ body -> isEffectNode body\n  LetRec _ _ body -> isEffectNode body\n  _ -> false\n\nexecuteIfOpaque :: TcoExpr -> PhpExpr -> PhpExpr\nexecuteIfOpaque expr phpExpr =\n  if isEffectNode expr then phpExpr\n  else PhpCall (PhpRaw "phpurs_execute_effect") [ phpExpr ]`
);

codeGen = codeGen.replace(/translateExprImpl :: String -> Array String -> Map String String -> Map String String -> Maybe String -> Array String -> Boolean -> Int -> TcoExpr -> \{ stmts :: Array PhpStmt, expr :: PhpExpr, nextId :: Int \}\ntranslateExprImpl modNameStr recVars namedBound bound mbNamedVar loopCtx isTail nextId tcoExpr@\(TcoExpr _ syntax\) =/,
  `translateExprImpl_ :: String -> Array String -> Map String String -> Map String String -> Maybe String -> Array String -> Boolean -> Boolean -> Int -> TcoExpr -> { stmts :: Array PhpStmt, expr :: PhpExpr, nextId :: Int }\ntranslateExprImpl_ modNameStr recVars namedBound bound mbNamedVar loopCtx isTail inEffectBlock nextId tcoExpr@(TcoExpr _ syntax) =\n  let\n    isEff = isEffectNode tcoExpr\n  in\n    if isEff && not inEffectBlock then\n      let\n        res = translateExprImpl_ modNameStr recVars namedBound bound mbNamedVar loopCtx false true nextId tcoExpr\n        fvs = freeVars tcoExpr\n        mappedFvs = map (\\v -> fromMaybe v (Map.lookup v bound)) (Array.fromFoldable fvs)\n        useVars = Array.nub (map (\\mapped -> if Array.elem mapped recVars then "&" <> mapped else mapped) mappedFvs)\n      in\n        { stmts: [], expr: PhpFunction useVars [] "" (res.stmts <> [ PhpReturn res.expr ]), nextId: res.nextId }\n    else`
);

codeGen = codeGen.replace(/translateExprImpl /g, "translateExprImpl_ ");

codeGen = codeGen.replace(/translateExprImpl_ modNameStr recVars namedBound bound Nothing \[\] false ([a-zA-Z0-9_\.]+) expr/g, "translateExprImpl_ modNameStr recVars namedBound bound Nothing [] false false $1 expr");

codeGen = codeGen.replace(/translateExprImpl_ modNameStr recVars namedBound bound \(Just v\) loopCtx isTail ([a-zA-Z0-9_\.]+) body/g, "translateExprImpl_ modNameStr recVars namedBound bound (Just v) loopCtx isTail false $1 body");

codeGen = codeGen.replace(/translateExprImpl_ modNameStr recVars namedBound bound \(Just varName\) \[\] false ([a-zA-Z0-9_\.]+) val/g, "translateExprImpl_ modNameStr recVars namedBound bound (Just varName) [] false false $1 val");

codeGen = codeGen.replace(/translateExprImpl_ modNameStr recVars namedBound newBound Nothing loopCtx isTail ([a-zA-Z0-9_\.]+) body/g, "translateExprImpl_ modNameStr recVars namedBound newBound Nothing loopCtx isTail false $1 body");

codeGen = codeGen.replace(/translateExprImpl_ modNameStr recVars namedBound newBound \(Just varName\) \[\] true ([a-zA-Z0-9_\.]+) innerBody/g, "translateExprImpl_ modNameStr recVars namedBound newBound (Just varName) [] true false $1 innerBody");

codeGen = codeGen.replace(/translateExprImpl_ modNameStr recVars namedBound acc\.bound Nothing loopCtx isTail ([a-zA-Z0-9_\.]+) body/g, "translateExprImpl_ modNameStr recVars namedBound acc.bound Nothing loopCtx isTail false $1 body");

codeGen = codeGen.replace(/translateExprImpl_ modNameStr recVars namedBound bound Nothing loopCtx isTail ([a-zA-Z0-9_\.]+) def/g, "translateExprImpl_ modNameStr recVars namedBound bound Nothing loopCtx isTail false $1 def");

codeGen = codeGen.replace(/translateExprImpl_ modNameStr recVars namedBound bound Nothing \[\] false ([a-zA-Z0-9_\.]+) condExpr/g, "translateExprImpl_ modNameStr recVars namedBound bound Nothing [] false false $1 condExpr");

codeGen = codeGen.replace(/translateExprImpl_ modNameStr recVars namedBound bound Nothing loopCtx isTail ([a-zA-Z0-9_\.]+) bodyExpr/g, "translateExprImpl_ modNameStr recVars namedBound bound Nothing loopCtx isTail false $1 bodyExpr");


fs.writeFileSync('../phpurs/phpurs/src/Phpurs/CodeGen.purs', codeGen);

File: patch.py
import re

with open('../phpurs/phpurs/src/Phpurs/CodeGen.purs', 'r') as f:
    content = f.read()

# Add isEffectNode and executeIfOpaque
content = content.replace(
    'freeVars :: TcoExpr -> Set String\nfreeVars (TcoExpr (TcoAnalysis { freeVars: fvs }) _) = fvs',
    '''freeVars :: TcoExpr -> Set String
freeVars (TcoExpr (TcoAnalysis { freeVars: fvs }) _) = fvs

isEffectNode :: TcoExpr -> Boolean
isEffectNode (TcoExpr _ syntax) = case syntax of
  EffectBind _ _ _ _ -> true
  EffectPure _ -> true
  EffectDefer _ -> false
  PrimEffect _ -> true
  UncurriedEffectApp _ _ -> true
  Let _ _ _ body -> isEffectNode body
  LetRec _ _ body -> isEffectNode body
  _ -> false

executeIfOpaque :: TcoExpr -> PhpExpr -> PhpExpr
executeIfOpaque expr phpExpr =
  if isEffectNode expr then phpExpr
  else PhpCall (PhpRaw "phpurs_execute_effect") [ phpExpr ]'''
)

# Replace function signature and body start
content = content.replace(
    '''translateExprImpl :: String -> Array String -> Map String String -> Map String String -> Maybe String -> Array String -> Boolean -> Int -> TcoExpr -> { stmts :: Array PhpStmt, expr :: PhpExpr, nextId :: Int }
translateExprImpl modNameStr recVars namedBound bound mbNamedVar loopCtx isTail nextId tcoExpr@(TcoExpr _ syntax) =
  let
    doTrace f =''',
    '''translateExprImpl_ :: String -> Array String -> Map String String -> Map String String -> Maybe String -> Array String -> Boolean -> Boolean -> Int -> TcoExpr -> { stmts :: Array PhpStmt, expr :: PhpExpr, nextId :: Int }
translateExprImpl_ modNameStr recVars namedBound bound mbNamedVar loopCtx isTail inEffectBlock nextId tcoExpr@(TcoExpr _ syntax) =
  let
    isEff = isEffectNode tcoExpr
  in
    if isEff && not inEffectBlock then
      let
        res = translateExprImpl_ modNameStr recVars namedBound bound mbNamedVar loopCtx false true nextId tcoExpr
        fvs = freeVars tcoExpr
        mappedFvs = map (\\v -> fromMaybe v (Map.lookup v bound)) (Array.fromFoldable fvs)
        useVars = Array.nub (map (\\mapped -> if Array.elem mapped recVars then "&" <> mapped else mapped) mappedFvs)
      in
        { stmts: [], expr: PhpFunction useVars [] "" (res.stmts <> [ PhpReturn res.expr ]), nextId: res.nextId }
    else
  let
    doTrace f ='''
)

def replacer(m):
    return f'translateExprImpl_ {m.group(1)} false {m.group(2)}'

content = re.sub(r'translateExprImpl (modNameStr.*?(?:\[\]|loopCtx|Nothing).*?(?:false|true|isTail)) ([a-zA-Z0-9_\.\+ \(\)]+) (expr|a|e|body|val|innerBody|def|condExpr|bodyExpr|fn|arg|argExpr|flatFn|e1|e2)', replacer, content)

content = content.replace(
    '''EffectBind (Just (Ident i)) (Level l) val body ->
    let
      oldVarName = localId (Just (Ident i)) (Level l)
      varName = oldVarName <> "_" <> show nextId
      resVal = translateExprImpl_ modNameStr recVars namedBound bound (Just varName) [] false false nextId val
      newBound = Map.insert oldVarName varName bound
      resBody = translateExprImpl_ modNameStr recVars namedBound newBound Nothing loopCtx isTail false (resVal.nextId + 1) body
    in
      { stmts: resVal.stmts <> [ PhpAssign varName resVal.expr ] <> resBody.stmts, expr: resBody.expr, nextId: resBody.nextId }''',
    '''EffectBind (Just (Ident i)) (Level l) val body ->
    let
      oldVarName = localId (Just (Ident i)) (Level l)
      varName = oldVarName <> "_" <> show nextId
      resVal = translateExprImpl_ modNameStr recVars namedBound bound (Just varName) [] false true nextId val
      newBound = Map.insert oldVarName varName bound
      resBody = translateExprImpl_ modNameStr recVars namedBound newBound Nothing loopCtx isTail true (resVal.nextId + 1) body
      valExpr = executeIfOpaque val resVal.expr
    in
      { stmts: resVal.stmts <> [ PhpAssign varName valExpr ] <> resBody.stmts, expr: resBody.expr, nextId: resBody.nextId }'''
)

content = content.replace(
    '''EffectBind Nothing (Level l) val body ->
    let
      oldVarName = localId Nothing (Level l)
      varName = oldVarName <> "_" <> show nextId
      resVal = translateExprImpl_ modNameStr recVars namedBound bound (Just varName) [] false false nextId val
      newBound = Map.insert oldVarName varName bound
      resBody = translateExprImpl_ modNameStr recVars namedBound newBound Nothing loopCtx isTail false (resVal.nextId + 1) body
    in
      { stmts: resVal.stmts <> [ PhpAssign varName resVal.expr ] <> resBody.stmts, expr: resBody.expr, nextId: resBody.nextId }''',
    '''EffectBind Nothing (Level l) val body ->
    let
      oldVarName = localId Nothing (Level l)
      varName = oldVarName <> "_" <> show nextId
      resVal = translateExprImpl_ modNameStr recVars namedBound bound (Just varName) [] false true nextId val
      newBound = Map.insert oldVarName varName bound
      resBody = translateExprImpl_ modNameStr recVars namedBound newBound Nothing loopCtx isTail true (resVal.nextId + 1) body
      valExpr = executeIfOpaque val resVal.expr
    in
      { stmts: resVal.stmts <> [ PhpAssign varName valExpr ] <> resBody.stmts, expr: resBody.expr, nextId: resBody.nextId }'''
)

content = content.replace(
    '''EffectDefer e ->
    let
      res = translateExprImpl_ modNameStr recVars namedBound bound Nothing [] false false nextId e''',
    '''EffectDefer e ->
    let
      res = translateExprImpl_ modNameStr recVars namedBound bound Nothing [] false true nextId e'''
)

with open('../phpurs/phpurs/src/Phpurs/CodeGen.purs', 'w') as f:
    f.write(content)
